quinn-proto/src/token_reuse_preventer.rs

```diff
@@ -0,0 +1,218 @@
+//! Limiting clients' ability to reuse tokens from NEW_TOKEN frames
+
+use std::{
+    hash::{Hash as _, Hasher as _},
+    mem::{size_of, swap},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use rustc_hash::{FxHashSet, FxHasher};
+use tracing::warn;
+
+/// Error for when a token may have been reused
+pub struct TokenReuseError;
+
+/// Responsible for limiting clients' ability to reuse tokens from NEW_TOKEN frames
+///
+/// [_RFC 9000 § 8.1.4:_](https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1.4)
+///
+/// > Attackers could replay tokens to use servers as amplifiers in DDoS attacks. To protect
+/// > against such attacks, servers MUST ensure that replay of tokens is prevented or limited.
+/// > Servers SHOULD ensure that tokens sent in Retry packets are only accepted for a short time,
+/// > as they are returned immediately by clients. Tokens that are provided in NEW_TOKEN frames
+/// > (Section 19.7) need to be valid for longer but SHOULD NOT be accepted multiple times.
+/// > Servers are encouraged to allow tokens to be used only once, if possible; tokens MAY include
+/// > additional information about clients to further narrow applicability or reuse.
+pub trait TokenReusePreventer: Send + Sync {
+    /// Called when a client uses a token from a NEW_TOKEN frame
+    ///
+    /// False negatives and false positives are both permissible.
+    fn using(
+        &mut self,
+        token_rand: u128,
+        issued: SystemTime,
+        new_token_lifetime: Duration,
+    ) -> Result<(), TokenReuseError>;
+}
+
+/// Bloom filter-based `TokenReusePreventer`
+///
+/// Parameterizable over an approximate maximum number of bytes to allocate. Starts out by storing
+/// used tokens in a hash set. Once the hash set becomes too large, converts it to a bloom filter.
+///
+/// Divides time into periods based on `new_token_lifetime` and stores two filters at any given
+/// moment, for each of the two periods currently non-expired tokens could expire in. As such,
+/// turns over filters as time goes on to avoid bloom filter false positive rate increasing
+/// infinitely over time.
+pub struct BloomTokenReusePreventer {
```

Comment:
Should we pull in a mature third-party bloom filter rather than rolling out own? A few moments' search turns up e.g. https://github.com/tomtomwombat/fastbloom/.

Reply:
To be clear, this is not a rhetorical question, I'm interested in opinions.

Reply:
External dependencies aren't free, but neither is vendoring complex logic, especially when we're not experts on it.

Reply:
Though might be nice to make it optional.

Reply:
I made it an optional dependency :)